### PR TITLE
[RFC] initial preferences API

### PR DESCRIPTION
**Pull request by qmx**

### CODE REVIEW

File: uberfire-api/src/main/java/org/uberfire/preferences/impl/DefaultResolutionStrategy.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import org.uberfire.preferences.ResolutionStrategy;
+import org.uberfire.preferences.Scope;
+
+@ApplicationScoped
+public class DefaultResolutionStrategy implements ResolutionStrategy {
+
+    @Override
+    public List<Scope> order( final Scope scope,
+                              final Scope... scopes ) {
+        return null;
```
Comment (porcelli) [inline uberfire-api/src/main/java/org/uberfire/preferences/impl/DefaultResolutionStrategy.java]: ?

**Comment by porcelli** [comment]
Missing a reasonable test suite.
Also missing the client side bits.. that i can help on it... 30 mins is done.

**Comment by porcelli** [comment]
Conversion tool from legacy to new format (now including a version info please).. would be very useful ;)

File: uberfire-api/src/main/java/org/uberfire/preferences/AsyncDictionary.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package org.uberfire.preferences;
+
+import org.uberfire.mvp.ParameterizedCommand;
+
+public interface AsyncDictionary {
+
+    <T> void get( final String key,
+                  final Class<T> clazz,
```
Comment (csadilek) [inline uberfire-api/src/main/java/org/uberfire/preferences/AsyncDictionary.java]: The parameter clazz doesn't seem to be used in the two implementations below.

**Comment by csadilek** [comment]
Do we have a JIRA for this? If not, let's create one and reference it in the commit message, please. Just picking some low-hanging fruit....been a long day here ;).

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/preferences/PreferenceStorageServiceBackendImpl.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    @Override
+    public void delete( final String key ) {
+
+    }
+
+    private Path buildStoragePath( final Scope scope,
+                                   final String key ) {
+
+        final String path;
+        if ( scope.equals( DefaultScopeTypes.USER.toScope() ) ) {
```
Comment (manstis) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/preferences/PreferenceStorageServiceBackendImpl.java]: Would it not be better to ask the ```Scope``` to build a "prefix"? e.g. ```path = scope.getPrefix()```

File: uberfire-api/src/main/java/org/uberfire/preferences/impl/PreferenceStoreImpl.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        return defaultScope;
+    }
+
+    @Override
+    public ResolutionStrategy resolutionStrategy() {
+        return resolutionStrategy;
+    }
+
+    @Override
+    public void put( String key,
```
Comment (manstis) [inline uberfire-api/src/main/java/org/uberfire/preferences/impl/PreferenceStoreImpl.java]: How would we put/get preferences to/from different scopes? It looks like we'd always be using ```defaultScope```?

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/preferences/PreferenceStorageServiceBackendImpl.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    @Named("systemFS")
+    private FileSystem fileSystem;
+
+    private final XStream xs = new XStream();
+
+    @Override
+    public <T> T read( final Scope scope,
+                       final String key,
+                       final ResolutionStrategy resolutionStrategy ) {
+        List<Scope> scopes = resolutionStrategy.order( scope );
```
Comment (manstis) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/preferences/PreferenceStorageServiceBackendImpl.java]: Based on @porcelli comment for ```DefaultResolutionStrategy``` and the ```@Produces``` you have in ```ApplicationScopedProducer``` won't this throw NPEs?

File: uberfire-api/src/main/java/org/uberfire/preferences/AsyncDictionary.java (hunk 2 above)
Comment (porcelli) [inline uberfire-api/src/main/java/org/uberfire/preferences/AsyncDictionary.java]: true that. the reason for that is basically that the default portable impl can't use clazz.cast as it won't compile on GWT. It should have a server side extension of it to take advantage of it.

File: uberfire-api/src/main/java/org/uberfire/preferences/impl/PreferenceStoreImpl.java (hunk 4 above)
Comment (porcelli) [inline uberfire-api/src/main/java/org/uberfire/preferences/impl/PreferenceStoreImpl.java]: AFAIK you'd use a forScope and from there you get what you're looking for. AND you'll also have the resolution strategy to define precedence.

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/preferences/PreferenceStorageServiceBackendImpl.java (hunk 3 above)
Comment (porcelli) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/preferences/PreferenceStorageServiceBackendImpl.java]: this is what scope.key is for. the only exception is User scope as it has a special behavior (global state)

File: uberfire-api/src/main/java/org/uberfire/preferences/AsyncDictionary.java (hunk 2 above)
Comment (csadilek) [inline uberfire-api/src/main/java/org/uberfire/preferences/AsyncDictionary.java]: ok, then I guess it should be factored out on the server, but I think that's what you meant.

**Comment by csadilek** [comment]
@qmx Can this (RFC) PR be closed assuming you will be sending a separate one for the feature?

**Comment by manstis** [comment]
What's happening with a consolidated preferences service now? Is it dead?
It'd be a great shame - we really need one in our platform.

**Comment by csadilek** [comment]
I agree with @manstis. Would really be a much needed feature.

**Pull request closed without merging**